### ISSUE TITLE
Camera Objects

### ISSUE THREAD

**Issue by topherbuckley** [mistic100/Photo-Sphere-Viewer#125]
I'm trying to link up [omnitone](https://github.com/GoogleChrome/omnitone) with Photo-Sphere-Viewer and the hurdle I'm working on at the moment is getting access to the camera object's matrix object in order to control/link the omnitone rotator to the camera. 

I noticed in the API that there is a getPosition() getter, but this only returns long and lat, which won't allow me to pair up with the omnitone rotator. I cloned and built your repo and went through the example/index.htm and can grab the right matrix while setting a break point after it has been defined on line 976 of photo-sphere-viewer.js, but I'm not sure how to access this after this initialization/prototype creation finishes. I'm not sure if this is accessible or not, and I am no JS veteran, so any push in the right direction would be greatly appreciated.

**Comment by mistic100** [comment]
With latitude and longitude you should be able to do what you need, it's just a bunch of trigonometrics formulas to get the x/y/z

All internal objects in PSV are actually public, just use ` PSV.camera` to get the THREE.PerspectiveCamera object.

**Comment by mistic100** [comment]
Would you notify me what you have a working demo of this ? It sounds (pun indented) awesome.

**Comment by topherbuckley** [comment]
Perfect! PSV.camera was exactly what I was looking for, thanks. I figured it could be done with some trig from the long/lat, but I didn't want to mess with all of that if I didn't have to. Also, not that I think I'll be using it with PSV anytime soon since your camera remains at [0,0,0] right? But for other applications, where the camera isn't fixed at the center, I thought having the 4x4 matrix could prove useful for other applications when used with the Web Audio API for distance attenuation type applications. Not sure exactly how that would work yet, but I'd imagine pairing PSV with something like what [Matterport](https://matterport.com/) is doing...but even there, one might just want to move everything around the origin rather than move the camera object itself. 

Sure, I'll let you know when I've finished a demo. I'm guessing you've already taken a look, but omnitone has its own "demo-player", but I couldn't find this anywhere opensource, and gave up trying to figure it out after running into too many minified sections of code. See their demos [here](https://googlechrome.github.io/omnitone/#gallery) if interested.

**Comment by topherbuckley** [comment]
I started trying to make this and ran into the same hurdle as before. In your example/index.htm I am able to access PSV.camera, but in my own implementation, which I detailed in my recent reply [here](https://github.com/mistic100/Photo-Sphere-Viewer/issues/126), there is no PSV object, and the object I created PhotoSphereViewer has no camera object attached it seems. The normal functionality of the Photo-Sphere-Viewer code, works just fine, but I'm just unable to access the camera object in this way. 

Any recommendations as to how to set up my PhotoSphereViewer object such that I can grab the camera object from it?

**Comment by mistic100** [comment]
Because the camera is not immediately created.

```js
viewer.on('ready', function() {
  var camera = viewer.camera;
});
```

**Comment by topherbuckley** [comment]
Alright. I got it. That wasn't exactly my problem, but you pointed me in the right direction. I just needed to assign an instance name when I initialized the PhotoSphereViewer class. 

```javascript
viewer = new PhotoSphereViewer({ /* ... */ })
```
Then I was able to reference it just fine. 

I was previously only doing:
```javascript
new PhotoSphereViewer({ /* ... */ })
```
Not sure if the later is truly refereanceable, but PhotoSphereViewer was accessible while the camera object was not. Now the viewer instance has a camera attached. 

Thanks!

**Comment by topherbuckley** [comment]
Huh, I ran into the problem you seemed to have anticipated. I added the 

```javascript
viewer.on('ready', function() {
  var camera = viewer.camera;
});
```
to my code, and set a breakpoint on the var camera ... line. 
The debugger never stopped, making me believe this ready on method isn't working quite right. I used the on render with the same code:


```javascript
viewer.on('render', function() {
  var camera = viewer.camera;
});
```

and the debugger stops within just fine. Not sure if thats something wrong on my end, or something buggy on your end. Is the on method something you made or something from threeJS? 

Anyways, I got PSV and omnitone to work together! Its alive! There are a few bugs to work out still, but I'll update you here after I have something nicer to look at and listen to.

Cheers!
